### ISSUE TITLE
Unrecognized GCOV output

### ISSUE THREAD

**Issue by nmoinvaz** [gcovr/gcovr#583]
**Describe the bug**
We have been experiencing parsing errors in our CI pipeline that utilizes gcovr for python for the past day.  This had not been happening for at least a few months prior since we incorporated gcovr.

**To Reproduce**
https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true
```
(WARNING) Unrecognized GCOV output for /home/runner/work/zlib-ng/zlib-ng/deflate_medium.c
	  branch  3 taken -7476[72](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:72)
	This is indicative of a gcov output parse error.
	Please report this to the gcovr developers
	at <https://github.com/gcovr/gcovr/issues>.
(WARNING) Exception during parsing:
	UnknownLineType: branch  3 taken -[74](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:74)[76](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:76)72
(ERROR) Exiting because of parse errors.
	You can run gcovr with --gcov-ignore-parse-errors
	to continue anyway.
...
(DEBUG)   Filtering coverage data for file /usr/include/x86_64-linux-gnu/bits/string_fortified.h
Traceback (most recent call last):
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 77, in worker
    work(*args, **kwargs)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 317, in process_datafile
    done = run_gcov_and_process_files(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 490, in run_gcov_and_process_files
    process_gcov_data(fname, covdata, abs_filename, options)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line [145](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:145), in process_gcov_data
    coverage = parse_coverage(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 370, in parse_coverage
    _report_lines_with_errors(lines_with_errors, context)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 509, in _report_lines_with_errors
    raise errors[0]  # guaranteed to have at least one exception
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 322, in parse_coverage
    tokenized_lines.append((_parse_line(raw_line), raw_line))
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 777, in _parse_line
    raise UnknownLineType(line)
gcovr.gcov_parser.UnknownLineType: branch  3 taken -747672
Traceback (most recent call last):
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 77, in worker
    work(*args, **kwargs)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 317, in process_datafile
    done = run_gcov_and_process_files(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 490, in run_gcov_and_process_files
    process_gcov_data(fname, covdata, abs_filename, options)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 145, in process_gcov_data
    coverage = parse_coverage(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 370, in parse_coverage
    _report_lines_with_errors(lines_with_errors, context)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 509, in _report_lines_with_errors
    raise errors[0]  # guaranteed to have at least one exception
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 322, in parse_coverage
    tokenized_lines.append((_parse_line(raw_line), raw_line))
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 777, in _parse_line
    raise UnknownLineType(line)
gcovr.gcov_parser.UnknownLineType: branch  3 taken -747672
(ERROR) Uncaught EXCEPTION
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/__main__.py", line 521, in <module>
    main()
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/__main__.py", line 308, in main
    collect_coverage_from_gcov(covdata, options)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/__main__.py", line 367, in collect_coverage_from_gcov
    contexts = pool.wait()
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 170, in __exit__
    self.wait()
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line [161](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:161), in wait
    raise self.exceptions[0][1]
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/__main__.py", line 367, in collect_coverage_from_gcov
    contexts = pool.wait()
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 161, in wait
    raise self.exceptions[0][1]
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 77, in worker
    work(*args, **kwargs)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 317, in process_datafile
    done = run_gcov_and_process_files(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 490, in run_gcov_and_process_files
    process_gcov_data(fname, covdata, abs_filename, options)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 145, in process_gcov_data
    coverage = parse_coverage(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 370, in parse_coverage
    _report_lines_with_errors(lines_with_errors, context)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 509, in _report_lines_with_errors
    raise errors[0]  # guaranteed to have at least one exception
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 322, in parse_coverage
    tokenized_lines.append((_parse_line(raw_line), raw_line))
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 777, in _parse_line
    raise UnknownLineType(line)
gcovr.gcov_parser.UnknownLineType: branch  3 taken -747672
```

**Expected behavior**
No parser errors.

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - GCC version 9.3.0
 - GCOVR version 5.1

**Comment by nmoinvaz** [comment]
Reverting to version 5.0 solves the gcovr parser errors.

**Comment by Spacetown** [comment]
The line in the gcov file is e.g. `branch  1 taken -2704160` but the count is negativ.

Looking at the latest sources the counter is `int64_t`. The source is https://github.com/gcc-mirror/gcc/blob/50bc9185c2821350f0b785d6e23a6e9dcde58466/gcc/gcov.c#L2861 where the formating is done in https://github.com/gcc-mirror/gcc/blob/50bc9185c2821350f0b785d6e23a6e9dcde58466/gcc/gcov.c#L2401 with `decimal_places=-1` which results in a call of https://github.com/gcc-mirror/gcc/blob/50bc9185c2821350f0b785d6e23a6e9dcde58466/gcc/gcov.c#L2370

Can this be a overflow in the counters of gcov? Or is the version of gcov other than gcc? Gcc is called with the full path `/usr/bin/gcc` and `gcov` without path.

Can you try to set the option run `/usr/bin/gcc --version` and `/usr/bin/gcov --version` in your build to check if the version matches?

**Comment by nmoinvaz** [comment]
https://github.com/nmoinvaz/zlib-ng/runs/5711622569?check_suite_focus=true
Versions of gcc and gcov match = 9.3.0.

**Comment by Spacetown** [comment]
@latk For me the only cause of this line can be a overflow in the counter but the counter is 64bit.

@nmoinvaz Can you modify the following regex to accept also a minus at the start and can check if this negative counters can result from an overflow?
https://github.com/gcovr/gcovr/blob/bd0804ce61a9489a3834caa58c1dad1201e8e278/gcovr/gcov_parser.py#L81

**Comment by nmoinvaz** [comment]
@Spacetown, would I fork this gcovr repository, edit that line, and then have the CI install my forked repo? Is there anything else I would need to do for what you are wanting? Should I have it print out any info?

**Comment by latk** [comment]
@Spacetown I agree with your analysis. An overflow in an `int64_t` is extremely unlikely. Clearly, the `arc->count` does contain a negative value at that point, but a bug is more likely than 2^63 traversals of a branch in the program's control flow graph.

The quoted line in gcov_parser.py is *technically* wrong because the gcov.c code *could* produce negative counts, but that is clearly a logic error. That gcovr 5.0 used to accept this kind of input just means that gcovr 5.0 used to silently accept corrupted data, though TBH it doesn't matter in this case because gcovr 5.0 would have counted that branch as covered. So my gut reaction regarding the regex is “wontfix”.

I'm assuming that this is either an upstream problem with gcc/gcov, or that the coverage data itself has been corrupted somehow. For example, this might occur if the code being tested has a memory safety violation such as a buffer overflow in a statically allocated variable. But, such a memory safety issue might just as well be present in the gcc/gcov code since they write very weird C.

Supporting the corruption hypothesis, the negative values seem to appear in different files and with different values in the two provided runs.

* Run 1:
  * file: `/home/runner/work/zlib-ng/zlib-ng/deflate_medium.c` line: `branch  3 taken -747672`
* Run 2:
  * file: `/home/runner/work/zlib-ng/zlib-ng/deflate_rle.c` line: `branch  3 taken -1959`
  * file: `/home/runner/work/zlib-ng/zlib-ng/arch/x86/compare256_avx2.c` line: `branch  1 taken -645379 (fallthrough)`

It would be interesting to know if this issue persists with more recent GCC versions.

**Comment by nmoinvaz** [comment]
@latk the problem also exists with GCC 10.3.0 https://github.com/nmoinvaz/zlib-ng/runs/5712450273?check_suite_focus=true.

**Comment by nmoinvaz** [comment]
Problem exists with GCC 11.1.0 https://github.com/nmoinvaz/zlib-ng/runs/5712490433?check_suite_focus=true using gcc-11 and gcov-11.

**Comment by latk** [comment]
That the problem persists across different compiler versions is weak evidence pointing towards the data in the .gcda files already being corrupted, which probably means that the coverage counters are already corrupted in-memory of the program being covered.

In any case, I don't think this is something that can be solved in gcovr. While gcovr alerted you to this corruption, it is not the cause. You can use the `gcovr --gcov-ignore-parse-errors` option to ignore these errors and get reports with the potentially-corrupted data, with the caveat that this would cause the `branch 3 taken -1959` line to be ignored entirely instead of reporting a covered branch. This should get your CI pipeline back running until you can find the cause of the problem.

**Comment by latk** [comment]
I've noticed that lcov recently introduced similar handling for negative coverage counters: they will emit a warning and set it to zero, roughly comparable but more fine-grained to our `--gcov-ignore-parse-errors` option.

* I first saw it in this LCOV commit https://github.com/linux-test-project/lcov/commit/238ea7c33ca0b96654ee9f4b5c770792954a40f3#diff-94da12f7252946090b66b51808c3912f35a9a05be7a4a5640f58c3322e14a2a1R1866-R1874
* see also LCOV pull request https://github.com/linux-test-project/lcov/pull/173

The PR links to a GCC bug (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68080) claiming that is was fixed in 2016 by introducing the `-fprofile-update=atomic` compiler flag. Without that option, counts from multithreaded programs will allegedly be corrupted. I don't think that's the full story though. In my testing, GCC 7.x and later use atomic updates automatically whenever `-pthread` is used, and I think some of the examples given here didn't even involve any concurrency/multithreading.
